roachprod: prepare library for roachprod-centralized

golgeek opened the pull request:
This patch prepares the roachprod library for use by roachprod-centralized.

The change has been split over 7 commits:
1. enrich vm.Provider interface:
  This commit adds an `IsLocalProvider()` and a `String()` method to the `vm.Provider` interface.
  `IsLocalProvider()` will be used to differentiate remote providers (served by the centralized API) from the local ones.
  `String()` allows to create a human-readable (and map key) representation of the provider in its current configuration (e.g. `{ProviderName}-{project}`, `{ProviderName}-{AccountNumber}`, ...)  

2. providers configs:
  This patch refactorizes the different provider to ease instantiating a provider with a config.
  In case of the `GCE` provider and the DNS provider, it also ensures that all options are stored in the provider struct to allow multiple provider instances acting with different configurations in parallel.

3. manage AWS SST creds in the AWS provider:
  This patch implements the AWS temporary credentials mechanism in the AWS provider to remove the need for globally defined environment variables. When the SST configuration is enabled, each `runCommand()` call now requests a short lived credential to the provider instance, and makes sure to pass it to the `awscli` command in its environment variables directly instead of relying on globally set environment variables.
  This allows executing two different AWS providers on two different accounts in parallel instead of running them in sequence like done during the garbage collection. The garbage collector mechanism will be rewritten to this new way in a follow-up patch.

4. split cloud.Cluster in its own type subpackage and refactor cloud.Cloud:
  In order for roachprod-centralized to use the roachprod library, and for the roachprod library to use the roachprod-centralized client, it is required to move the `cloud.Cluster` and `cloud.Clusters` types in their own package to avoid import cycles. These two types are moved under `cloud/types` and all imports are aliased as `cloudcluster` to ensure better clarity and readability in the code.
  This patch also refactors `cloud.Cloud` so that the `ListCloud` can be executed against a pre-defined set of providers instead of the global `vm.AllProviderNames()`. This is achieved by storing two providers slices (one for remote, one for local) in the `cloud.Cloud` type, and having a `ListCloud` method that runs against these providers. The public `ListCloud` helper function has been rewritten to init a new `cloud.Cloud` with the global providers and use the object's method to avoid duplicating code.

5. support dns provider on all clouds and enrich vm.DNSProvider interface:
  Prior to this patch, only the GCE cloud provider had support for VMs public DNS at creation/deletion time. All other Cloud Providers were relying on the `Sync` operation to create DNS entries for all synced clusters.
  This patch brings support for instantiating all Cloud Providers with a DNS Provider, and populates the required `vm.VM` fields to make DNS behave the same across all Cloud Providers. The vm.DNSProvider interface is extended to also contain the DNS methods of the InfraProvider interface, which will allow a better separation of concerns.

6. start using gcp sdk instead of gcloud
  Prior to this patch the GCE provider was 100% built around executing gcloud commands. While this works fine with the local roachprod binary,  it creates two major issues when using it as a library:                  
  - the output of gcloud commands can be huge and is cached by the kernel in the page cache and on top of this, Go has to reserve more or less the same memory allocations for json deserialization. The combination leads to memory exhaustion on low memory runners.                      
  - using the auto-magical `glcoud dns record-sets import` command can fail due to inconsistencies between the state when the bind file is created and when the command is executed if other systems update the DNS zone at the same time.                                             
  This patch introduces the use of the GCP Golang SDK to list instances (and instance templates) by patching the GCE provider, and manage DNS records by creating a new SDK based DNS provider implementation.         
  The introduction of the new SDK DNS provider also ensures a better independence between the GCE provider and the DNS provider by fully using an interface in the VM provider. This greatly improves stability in long running operations and improves the overall quality of the code.

7. respect context in vm.Provider list and dsn paths
  Prior to this patch, ListCloud() and SyncDNS() paths in the roachprod library did not provide a way to pass a context.Context. This was preventing roachprod-centralized to respect tasks timeouts.
  This patch adds a context in the new `cloud.Cloud.ListCloud()` function and a `SyncDNSWithContext()` function in the `vm.DNSProvider` task. All non-context versions call the context version by passing an argument `context.Background()`.
  With the exception of the `gcloud` version of the `gce` provider, the `provider.List()` path has been updated to respect the provided context.

Epic: none
Release note: None
Informs:

golgeek committed:
roachprod: prepare centralized - enrich vm.Provider interface

This patch prepares the ground work for roachprod-centralized by
enriching the vm.Provider interface with two new methods:
- `String()`: returns a human-readable (and map key) for the provider in
  its current configuration (e.g. `{ProviderName}-{project}`,
`{ProviderName}-{AccountNumber}`, ...)
- `IsCentralizedProvider()`: returns whether or not the provider runs
  from the centralized service or is executed locally

These two new methods will ease the work of spawning multiple instances
of the same provider, in different configurations.

Epic: none
Release note: None

golgeek committed:
roachprod: prepare centralized - providers configs

This patch refactorizes the different providers to bring:
- a uniform way of instantiating a provider
- a better isolation of providers configurations to allow multiple
  providers to be instantiated in parallel

Epic: none
Release note: None

golgeek committed:
roachprod: prepare centralized - manage AWS SST creds in the provider

This patch adds provider management for the AWS SST credentials.

This allows each provider instance to generate its own short lived
credentials and pass them as environment variables to the aws CLI
without impacting other provider instances.

Epic: none
Release note: None

golgeek committed:
roachprod: prepare centralized - split cloud.Cluster in its own package

In order for roachprod-centralized to use the roachprod library,
and for the roachprod library to use the roachprod-centralized client,
it is required to move the `cloud.Cluster` and `cloud.Clusters` types
in their own package to avoid import cycles. These two types are moved
under `cloud/types` and all imports are aliased as `cloudcluster` to
ensure clarity in the code.

This patch also refactors `cloud.Cloud` so that `ListCloud` can be
executed against a pre-defined set of providers instead of the global
`vm.AllProviderNames()`. This is achieved by storing two providers
slices (one for remote, one for local) in the `cloud.Cloud` type, and
having a `ListCloud` method that runs against these providers. The
public `ListCloud` helper function has been rewritten to init a new
`cloud.Cloud` with the global providers and use the object's method to
avoid duplicating code.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154386)
<!-- Reviewable:end -->

golgeek committed:
roachprod: prepare centralized - support dns provider on all clouds

Prior to this patch, only the GCE cloud provider had support for VMs
public DNS at creation/deletion time. All other Cloud Providers were
relying on the `Sync` operation to create DNS entries for all synced
clusters.

This patch brings support for instantiating all Cloud Providers with a
DNS Provider, and populates the required `vm.VM` fields to make DNS
behave the same across all Cloud Providers. The vm.DNSProvider interface
is extended to also contain the DNS methods of the InfraProvider
interface, which will allow a better separation of concerns.

Epic: none
Release note: None

golgeek committed:
roachprod: prepare centralized - start using gcp sdk

Prior to this patch the GCE provider was 100% built around executing
gcloud commands. While this works fine with the local roachprod binary,
it creates two major issues when using it as a library:
- the output of gcloud commands can be huge and is cached by the kernel
  in the page cache and on top of this, Go has to reserve more or less
  the same memory allocations for json deserialization. The combination
  leads to memory exhaustion on low memory runners.
- using the auto-magical `glcoud dns record-sets import` command can
  fail due to inconsistencies between the state when the bind file is
  created and when the command is executed if other systems update the
  DNS zone at the same time.

This patch introduces the use of the GCP Golang SDK to list instances
(and instance templates) by patching the GCE provider, and manage DNS
records by creating a new SDK based DNS provider implementation. The new
GCE SDK DNS provider is also named `gce` to be retro-compatible with the
`glcoud` based provider, as they have the exact same capabilities. The
`gcloud` DNS provider will be deprecated.

The introduction of the new SDK DNS provider also ensures a better
independence between the GCE provider and the DNS provider by fully
using an interface in the VM provider. This greatly improves stability
in long running operations and improves the overall quality of the code.

Epic: none
Release note: None

golgeek committed:
roachprod: prepare centralized - respect context in list and dns paths

Prior to this patch, ListCloud() and SyncDNS() paths in the roachprod
library did not provide a way to pass a context.Context. This was
preventing roachprod-centralized to respect tasks timeouts.

This patch adds a context in the new `cloud.Cloud.ListCloud()` function
and a `SyncDNSWithContext()` function in the `vm.DNSProvider` task. All
non-context versions call the context version by passing an argument
`context.Background()`.

With the exception of the `gcloud` version of the `gce` provider, the
`provider.List()` path has been updated to respect the provided context.

Epic: none
Release note: None

DarrylWong commented on pkg/roachprod/vm/aws/aws.go:
```diff
@@ -66,8 +66,14 @@ func Init() error {
+			&Provider{},
+			fmt.Sprintf("unable to init aws provider: %s", err),
+		)
+		return errors.New("doesn't have the required version")
```
Shouldn't we return the original error here? 
e.g. `return errors.Wrap(err, "unable to init aws provider")`

golgeek replied:
Sorry, quick copy/paste that went wrong. Fixing.

DarrylWong commented on pkg/roachprod/cloud/types/types.go:
```diff
@@ -0,0 +1,236 @@
+
+// A Cluster is created by querying various vm.Provider instances.
+//
+// TODO(benesch): unify with syncedCluster.
```
Is this (7+ year old) TODO still relevant? We probably don't have the context for what it fully means anymore, but it seems like after this change we _don't_ want to unify them anymore?

golgeek replied:
Indeed, probably not relevant anymore. Dropping it.

DarrylWong reviewed (commented):
>In order for roachprod-centralized to use the roachprod library,
and for the roachprod library to use the roachprod-centralized client

What's the expected dependencies between these three? Specifically why would the roachprod library need to use the roachprod centralized client?

golgeek commented:
> > In order for roachprod-centralized to use the roachprod library,
> > and for the roachprod library to use the roachprod-centralized client
> 
> What's the expected dependencies between these three? Specifically why would the roachprod library need to use the roachprod centralized client?

Best is probably to have a look at [this commit](https://github.com/cockroachdb/cockroach/pull/155012/commits/41acdfd7a788793392021f1b97b8df6c3f93d3b1) to undertand the logic (in PR #3, still a draft until the first two are merged).

Basically, since we want to keep support for running the `cloud.ListCloud()` locally, the less disruptive way is to inject a centralized hook in the local path.

We have two constraints:
1. we want to keep support for local sync (if centralized env var is not present)
2. we still need to sync the local clusters

These constraints make that the least disruptive way is to inject a roachprod-centralized hook in the local path:
- we start sync with the roachprod lib
- if local mode, we load all providers locally
- else, we load the remote providers from the roachprod-centralized client + we load the local providers locally

Might be easier explained with some ASCII art:

```
                                 roachprod list
                                       |
                                       ▼
                      roachprod/roachprod.go:List()
                                       |
                                       ▼
                  roachprod/cloud/cluster_cloud.go:ListCloud()
                       |                                 |
         a) local mode |                                 | b) centralized mode
                       |                                 |
                       ▼                                 ▼
            all providers.List()      roachprod-centralized/client.go:ListClusters()
                                                         +
                                               local providers.List()
```

Since both (local and roachprod-centralized) use and return the `cloud.Cluster{}` type for compatibility and simplicity, this was creating a cycle dependency, which was broken by moving it in its own sub-package.

Does it make sense?

DarrylWong reviewed (approved):
A few more comments, but I don't see anything wrong at a high level so LGTM.

williamchoe3 commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -119,30 +122,111 @@ func Init() error {
+				(&SDKDNSProviderOpts{}).NewFromGCEDNSProviderOpts(p.dnsProviderOpts),
+			)
+			if err != nil {
+				return nil, errors.Wrapf(err, "unable to init dns client")
```
not sure if the linter is gonna complain about `Wrapf()` without any additional args, if it doesn't then doesn't matter to me / can clean it up later

williamchoe3 commented:
Took a very high level look. Not something that needs to be addressed here, but just wondering on how to approach testing particularly the in `vm/*` and in the centralized service in general.

For unit testing the interface methods, I think the the only way to write a test for the provider methods would be to wrap the provider SDK calls in an interface and then create mocks which i don't know is worth it to just running some functional tests against a staging env eventually.

But i'd still like to increase the unit testing coverage if possible? Relying only on functional tests also isn't the best dev process. Curious if anyone has any thoughts on this or an easy way to increase unit test coverage?

srosenberg commented on pkg/roachprod/vm/flagstub/flagstub.go:
```diff
@@ -28,6 +29,11 @@ type provider struct {
 	unimplemented string
 }
 
+// IsLocalProvider returns true because it is a local provider.
```
This seems like a circular definition since it's a stub, neither local nor remote.

golgeek commented on pkg/roachprod/vm/vm.go:
```diff
@@ -556,6 +562,11 @@ type Provider interface {
 	CreateProviderOpts() ProviderOpts
 	CleanSSH(l *logger.Logger) error
 
+	// IsLocalProvider returns true if the provider is a local provider.
```
Maybe `RunsLocally() bool` would be less confusing?

Current design hardcodes `false` for cloud providers, but intent is that we could add a config on each provider to have this value dynamically set, and one could choose to run a cloud provider locally, i.e., not via the centralized service.

golgeek replied:
I really don't mind renaming to `IsCentralizedProvider()`, but I have a feeling the naming issue rests more on the `local` provider itself. We could have multiple providers that exec CRDB on the local machine (binary, docker, vagrant, ...).
